SDL/SDLJoystick.cpp

```diff
@@ -11,167 +11,209 @@ extern "C" {
+			}
+			else
+			{
+				printf("control pad mapping: %s\n", SDL_GameControllerMapping(player1Controller));
+			}
+		}
+		else
+		{
+			printf("Error! Could not open controller\n");
+		}
 	}
-
-        if (g_Config.bPS3Controller)
-                fillMappingPS3();
-        else
-                fillMapping();
-
-
 }
 
 SDLJoystick::~SDLJoystick(){
 	if (thread)
 	{
 		running = false;
+		if (player1Controller)
+		{
+			SDL_GameControllerClose(player1Controller);
+		}
 		SDL_Event evt;
 		evt.type = SDL_USEREVENT;
 		SDL_PushEvent(&evt);
 		SDL_WaitThread(thread,0);
 	}
-	for (SDL_Joystick *joy : joys)
-	{
-		SDL_JoystickClose(joy);
-	}
 }
 
 void SDLJoystick::startEventLoop(){
 	thread = SDL_CreateThread(SDLJoystickThreadWrapper, "joystick",static_cast<void *>(this));
 }
 
+keycode_t SDLJoystick::getKeycodeForButton(SDL_GameControllerButton button)
+{
+	switch (button)
+	{
+		case SDL_CONTROLLER_BUTTON_DPAD_UP: return NKCODE_DPAD_UP;
+		case SDL_CONTROLLER_BUTTON_DPAD_DOWN: return NKCODE_DPAD_DOWN;
+		case SDL_CONTROLLER_BUTTON_DPAD_LEFT: return NKCODE_DPAD_LEFT;
+		case SDL_CONTROLLER_BUTTON_DPAD_RIGHT: return NKCODE_DPAD_RIGHT;
+		case SDL_CONTROLLER_BUTTON_A: return NKCODE_BUTTON_2;
+		case SDL_CONTROLLER_BUTTON_B: return NKCODE_BUTTON_3;
+		case SDL_CONTROLLER_BUTTON_X: return NKCODE_BUTTON_4;
+		case SDL_CONTROLLER_BUTTON_Y: return NKCODE_BUTTON_1;
+		case SDL_CONTROLLER_BUTTON_RIGHTSHOULDER: return NKCODE_BUTTON_5;
+		case SDL_CONTROLLER_BUTTON_LEFTSHOULDER: return NKCODE_BUTTON_6;
+		case SDL_CONTROLLER_BUTTON_START: return NKCODE_BUTTON_10;
+		case SDL_CONTROLLER_BUTTON_BACK: return NKCODE_BUTTON_9; // select button
+		case SDL_CONTROLLER_BUTTON_GUIDE: return NKCODE_BACK; // pause menu
+	}
+	return NKCODE_UNKNOWN;
+}
+
 void SDLJoystick::ProcessInput(SDL_Event &event){
 	switch (event.type) {
-	case SDL_JOYAXISMOTION:
+		case SDL_CONTROLLERBUTTONDOWN:
 		{
-			std::map<int, int>::const_iterator i = SDLJoyAxisMap.find(event.jaxis.axis);
-			int deviceIndex = getDeviceIndex(event.jaxis.which);
-			if (i != SDLJoyAxisMap.end() && deviceIndex >= 0) {
-				AxisInput axis;
-				axis.axisId = i->second;
-				// 1.2 to try to approximate the PSP's clamped rectangular range.
-				axis.value = 1.2 * event.jaxis.value / 32767.0f;
-				if (axis.value > 1.0f) axis.value = 1.0f;
-				if (axis.value < -1.0f) axis.value = -1.0f;
-				axis.deviceId = DEVICE_ID_PAD_0 + deviceIndex;
-				axis.flags = 0;
-				NativeAxis(axis);
-			}
-			break;
-		}
-
-	case SDL_JOYBUTTONDOWN:
-		{
-			std::map<int, int>::const_iterator i = SDLJoyButtonMap.find(event.jbutton.button);
-			int deviceIndex = getDeviceIndex(event.jbutton.which);
-			if (i != SDLJoyButtonMap.end() && deviceIndex >= 0) {
-				KeyInput key;
-				key.flags = KEY_DOWN;
-				key.keyCode = i->second;
-				key.deviceId = DEVICE_ID_PAD_0 + deviceIndex;
-				NativeKey(key);
+			if (event.cbutton.which == player1JoyInstanceID && event.cbutton.state == SDL_PRESSED)
```

Comment:
On Linux at least, the device ID is not unique to the control pad. For example, if an XBOX control pad is plugged in, this will be assigned device ID 0. Then if this control pad is disconnected and another say PS3 control pad is connected its device ID will be 0.

SDL2 assigns its own GUID to each model of control pad which is how control pad mappings are handled. Of course, this is SDL's own system and may not translate well with other input systems used by PPSSPP.